utils/stall_free.hh

```diff
@@ -174,6 +176,14 @@ future<> clear_gently(foreign_ptr<T>& o) noexcept {
     });
 }
 
+template <typename... T>
+requires (std::is_rvalue_reference_v<T&&> && ...)
+future<> clear_gently(T&&... o) noexcept {
+    return do_with(std::move(o)..., [](auto&... args) {
```

Comment:
In practice our types are move-friendly. Good moves are quite critical for us.

clear_gently() is an established API, not sure why it's modified here.

Reply:
Looks at the patch context, I see it an it makes sense.

Reply:
> Maybe we can make the caller responsible for keeping the object alive until the call finishes. They are passing a reference. It's easy to do when done from coroutines for example.

We can't, this is the whole point of this overload, we are given with a bunch of to-be-destroyed stuff and promise to take care of it, including keeping it alive while destroying. 

> The best solution here IMO is not needing this function at all. Looking at the next commit, I think the previous solution with background_clear_holder was good, just needs to be extended with the new fields.

I don't like the ad-hoc struct and the artificially specific code for a problem that can be formulated once in general terms and easily reused later. `std::array` of large size in an extreme case, people usually don't pass such a heavy beasts as move parameters. We can rename this function to `clear_gently_efficiently_movable` as a mitigation.

Reply:
> > Maybe we can make the caller responsible for keeping the object alive until the call finishes. They are passing a reference. It's easy to do when done from coroutines for example.
> 
> We can't, this is the whole point of this overload, we are given with a bunch of to-be-destroyed stuff and promise to take care of it, including keeping it alive while destroying.

The traditional way to do this is to start a background fiber that eats some queue, have it wait on the queue and consume the queue. To destroy something in the background means pushing into the queue and waking the fiber. This avoid a (void)do_something_in_the_background() which always leads to problems. The background fiber can be collected with stop().


> 
> > The best solution here IMO is not needing this function at all. Looking at the next commit, I think the previous solution with background_clear_holder was good, just needs to be extended with the new fields.
> 
> I don't like the ad-hoc struct and the artificially specific code for a problem that can be formulated once in general terms and easily reused later. `std::array` of large size is an extreme case, people usually don't pass such heavy beasts as move parameters. We can rename this function to `clear_gently_efficiently_movable` as a mitigation.

I don't think it's necessary. Those arrays will cause problems everywhere they're passed as parameters. This isn't a framework like Seastar, we control the environment.

Reply:
> The traditional way to do this is to start a background fiber that eats some queue, have it wait on the queue and consume the queue. To destroy something in the background means pushing into the queue and waking the fiber. This avoid a (void)do_something_in_the_background() which always leads to problems. The background fiber can be collected with stop().

The code already does roughly that, except that background fiber accepts futures, not objects, since there are different types of objects to destroy - `unordered_map` and `boost::icl::interval_map`.

We could refactor it to something like this:

```
_factory->submit_background_work([_replication_map = std::move(_replication_map), ...]() {
    co_await clean_gently(_replication_map);
   // the same for other fields that need to be destroyed
});
```
don't think it's simpler.

Reply:
The new clear_gently is fine. I objected to throwing the future on the floor, but if you're not doing that, all is well.